client/widmgr/widmgr.go

```diff
@@ -4,80 +4,377 @@
 package widmgr
 
 import (
+	"context"
 	"fmt"
+	"sync"
+	"time"
 
+	"github.com/hashicorp/go-hclog"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/helper"
 	"github.com/hashicorp/nomad/nomad/structs"
 )
 
-type RPCer interface {
-	RPC(method string, args any, reply any) error
+// IdentityManager defines a manager responsible for signing and renewing
+// signed identities. At runtime it is implemented by *widmgr.WIDMgr.
+type IdentityManager interface {
+	Run() error
+	Get(cstructs.TaskIdentity) (*structs.SignedWorkloadIdentity, error)
+	Watch(cstructs.TaskIdentity) (<-chan *structs.SignedWorkloadIdentity, func())
+	Shutdown()
 }
 
-// Config wraps the configuration parameters the workload identity manager
-// needs.
-type Config struct {
-	// NodeSecret is the node's secret token
-	NodeSecret string
+type WIDMgr struct {
+	allocID  string
+	minIndex uint64
+	widSpecs map[string][]*structs.WorkloadIdentity // task -> WI
+	signer   IdentitySigner
 
-	// Region of the node
-	Region string
+	// lastToken are the last retrieved signed workload identifiers keyed by
+	// TaskIdentity
+	lastToken     map[cstructs.TaskIdentity]*structs.SignedWorkloadIdentity
+	lastTokenLock sync.RWMutex
 
-	RPC RPCer
-}
+	watchers     map[cstructs.TaskIdentity]chan *structs.SignedWorkloadIdentity
+	watchersLock sync.Mutex
 
-// WIDMgr fetches and validates workload identities.
-type WIDMgr struct {
-	nodeSecret string
-	region     string
-	rpc        RPCer
+	// minWait is the minimum amount of time to wait before renewing. Settable to
+	// ease testing.
+	minWait time.Duration
+
+	stopCtx context.Context
+	stop    context.CancelFunc
+
+	logger hclog.Logger
 }
 
-// New workload identity manager.
-func New(c Config) *WIDMgr {
+func NewWIDMgr(signer IdentitySigner, a *structs.Allocation, logger hclog.Logger) *WIDMgr {
+	widspecs := map[string][]*structs.WorkloadIdentity{}
+	tg := a.Job.LookupTaskGroup(a.TaskGroup)
+	for _, task := range tg.Tasks {
+		// Omit default identity as it does not expire
+		if len(task.Identities) > 0 {
+			widspecs[task.Name] = helper.CopySlice(task.Identities)
+		}
+	}
+
+	// Create a context for the renew loop. This context will be canceled when
+	// the allocation is stopped or agent is shutting down
+	stopCtx, stop := context.WithCancel(context.Background())
+
 	return &WIDMgr{
-		nodeSecret: c.NodeSecret,
-		region:     c.Region,
-		rpc:        c.RPC,
+		allocID:   a.ID,
+		minIndex:  a.CreateIndex,
+		widSpecs:  widspecs,
+		signer:    signer,
+		minWait:   10 * time.Second,
+		lastToken: map[cstructs.TaskIdentity]*structs.SignedWorkloadIdentity{},
+		watchers:  map[cstructs.TaskIdentity]chan *structs.SignedWorkloadIdentity{},
+		stopCtx:   stopCtx,
+		stop:      stop,
+		logger:    logger.Named("identity"),
```

Comment:
This logger's name is the same as the `identity_hook`. Should it have its own name like:

```suggestion
		logger:    logger.Named("widmgr"),
```